### DISCUSSION TITLE
Using a DB dependency in FastAPI without having to pass it through a function tree

### DISCUSSION THREAD

**Discussion by Maxpowerfreak** [fastapi/fastapi#6889] [Questions] [answered]
### First check

* [*] I added a very descriptive title to this issue.
* [*] I used the GitHub search to find a similar issue and didn't find it.
* [*] I searched the FastAPI documentation, with the integrated search.
* [*] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [*] I already read and followed all the tutorial in the docs and didn't find an answer.
* [*] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [*] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [*] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [*] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

This is not a working example, as this question is more for a complex app than a simple one. However, this is an example of what the app looks like structurally.

This app, when complete, will be working with a lot of models (50+), so queries/business logic is stored in services for each given models. Some of the more complex business logic has their own services, which call functions in other services. 

I'm skipping the main.py file, as it does not do anything fancy, simply adds the APIRouters and launches the app.

databases/mysql.py (one of 2 dbs in the project)
```python
from sqlalchemy import create_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker, Session

def get_uri():
    return 'the mysql uri'

engine = create_engine(get_uri())

SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)

Base = declarative_base()


def get_db():
    db: Session = SessionLocal()
    try:
        yield db
        db.commit()
    except Exception:
        db.rollback()
    finally:
        db.close()

```

controllers/controller1.py
```python
from fastapi import APIRouter, HTTPException, Path, Depends
from sqlalchemy.orm import Session
from services.mysql.bar import get_bar_by_id
from services.mysql.process_x import bar_process
from databases.mysql import get_db

router = APIRouter(prefix='/foo')


@router.get('/bar/{bar_id}')
def process_bar(bar_id: int = Path(..., title='The ID of the bar to process', ge=1),
                          mysql_session: Session = Depends(get_db)):
    # From the crontroller, to a service which only runs a query. This is fine.
    bar = get_bar_by_id(bar_id, mysql_session)

    if bar is None:
        raise HTTPException(status_code=404,
                            detail='Bar not found for id: {bar_id}'.format(bar_id=bar_id))

    # This one calls a function in a service which has a lot of business logic but no queries
    processed_bar = bar_process(bar, mysql_session)

    return processed_bar

```

services/mysql/process_x.py
```python
from .process_bar import process_the_bar
from models.mysql.w import W
from models.mysql.bar import Bar
from models.mysql.y import Y
from models.mysql.z import Z
from sqlalchemy.orm import Session


def w_process(w: W, mysql_session: Session):
    ...


def bar_process(bar: Bar, mysql_session: Session):
    # Very simplified, there's actually 5 conditional branching service calls here
    return process_the_bar(bar, mysql_session)


def y_process(y: Y, mysql_session: Session):
    ...


def z_process(z: Z, mysql_session: Session):
    ...

```

services/mysql/process_bar.py
```python
from . import model_service1
from . import model_service2
from . import model_service3
from . import additional_rules_service
from libraries.bar_functions import do_thing_to_bar
from models.mysql.bar import Bar
from sqlalchemy.orm import Session


def process_the_bar(bar: bar, mysql_session: Session):
    process_result = list()

    # Many processing steps, not all of them require db and might work on the bar directly
    process_result.append(process1(bar, mysql_session))
    process_result.append(process2(bar, mysql_session))
    process_result.append(process3(bar, mysql_session))
    process_result.append(process4(bar))
    process_result.append(...(bar))
    process_result.append(processY(bar))


def process1(bar: Bar, mysql_session: Session):
    return model_service1.do_something(bar.val, mysql_session)


def process2(bar: Bar, mysql_session: Session):
    return model_service2.do_something(bar.val, mysql_session)


def process3(bar: Bar, mysql_session: Session):
    return model_service3.do_something(bar.val, mysql_session)

def process4-Y(bar: Bar, mysql_session: Session):
    # do something using the bar library, or maybe on another service with no queries
    return list()
```

### Description

I am currently working on a POC using FastAPI on a bigger application. While everything works, I've gotten some push-back from some members of my team regarding the approach of handling database sessions through dependency injection of the Session through yield. The biggest issue being mainly having to pass the Session from the controller, to a service, to a second service and (in a few cases), a third service further in. In those cases, the intermediary service functions tend to have no database queries but the functions that they call on other services might have some. The complaint mainly lies in this being more difficult to maintain and having to pass the DB object everywhere seeming uselessly repetitive.

As you can see in my example above, we're stuck passing the mysql_session and having it repeat everywhere when using this approach.

Here are a two solutions I have thought of:

 1. Adding the DB session to the Starlette request state

I could do this either through the app.startup event ( https://fastapi.tiangolo.com/advanced/events/ ) or a middleware. However, it does mean passing the request state back and forth in a similar fashion (if my understanding of it is correct)

 2. Session scope approach using Context Manager

Pretty much, I would turn the get_db function into a context manager instead and not inject it as a dependency. By far the cleanest end result, however it goes completely against the concept of sharing a single db session across the request. 

I've considered the fully async approach using encode/databases as shown in the FastAPI documentation ( https://fastapi.tiangolo.com/advanced/async-sql-databases/ ), however one of the databases we are working with on SqlAlchemy is used through a plugin and I am assuming does not support async out of the box (Vertica). If I'm wrong, then I could consider the fully async approach.

So in the end, what I'm wondering is if it's possible to accomplish something "cleaner" without compromising the single session per request approach? And, is there a "FastAPI" way of doing so?

### Environment

* OS: Linux (Ubuntu 20.04)
* FastAPI Version: 0.63.0:
* sqlalchemy Version: 1.3.22
* Python version: 3.8.5

### Additional context

I'm new to professional Python, but not to programming. There might be a very obvious approach here that I am not seeing due to my limited work with the language so far (mainly Airflow). There is no one on my team with experience using FastAPI or Python for APIs overall, which is why I am asking for help here. I have also initially posted this question on StackOverflow, as I only recently learned that this was the better place to post questions.  https://stackoverflow.com/questions/66464098/using-a-db-dependency-in-fastapi-without-having-to-pass-it-through-a-function-tr

**Comment by insomnes** [comment]
@Maxpowerfreak hello!
May be i don't understand your case well, but you can encapsulate in separate BarProcessing class with data and session in it (or simple ds) and send it as argument to all of your functions.

**Comment by Maxpowerfreak** [comment]
@insomnes , thanks for the answer! That would be a decent solution, yes. I could apply something similar for all of the other bigger processes that we have and have the db object be saved as a variable on there too. It would be nice to have a global solution, if that's a possibility but this is a great idea for a fallback :)

What my other team members are expecting from this is for this to work something like Laravel or Flask (with flask-sqlalchemy) might, where you `import DB` when you have to do some queries instead of passing it as an argument through every function. 

However, when done as a dependency injection two things occur:
 1. One session per request
 2. One transaction per request (it only gets commit afterwards, as per the code for `get_db()`

What I'm wondering is if there's a way to achieve a similar result without going through a dependency injection, but instead only importing the db object when it is needed. Is declaring it on app boot something that is recommended? As in, appending it to the app object and importing it when it is needed?

**Comment by insomnes** [comment, marked as answer]
@Maxpowerfreak So, as far as i get it, you can create context var of DB session then:
https://docs.python.org/3/library/contextvars.html

```python3
from contextvars import ContextVar

from sqlalchemy import create_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker, Session

def get_uri():
    return 'the mysql uri'

engine = create_engine(get_uri())

SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)

Base = declarative_base()


def get_db():
    db: Session = SessionLocal()
    try:
        yield db
        db.commit()
    except Exception:
        db.rollback()
    finally:
        db.close()

db_session: ContextVar[Session] = ContextVar('db_session')

```

Then in your handlers:
```python3
from fastapi import APIRouter, HTTPException, Path, Depends
from sqlalchemy.orm import Session
from services.mysql.bar import get_bar_by_id
from services.mysql.process_x import bar_process
from databases.mysql import get_db, db_session

router = APIRouter(prefix='/foo')


@router.get('/bar/{bar_id}')
def process_bar(bar_id: int = Path(..., title='The ID of the bar to process', ge=1),
                          mysql_session: Session = Depends(get_db)):
    # From the crontroller, to a service which only runs a query. This is fine.
    db_session.set(mysql_session)
    bar = get_bar_by_id(bar_id)

    if bar is None:
        raise HTTPException(status_code=404,
                            detail='Bar not found for id: {bar_id}'.format(bar_id=bar_id))

    # This one calls a function in a service which has a lot of business logic but no queries
    processed_bar = bar_process(bar)

    return processed_bar
```


And in proccess functions:
```python3
from . import model_service1
from . import model_service2
from . import model_service3
from . import additional_rules_service
from databases.mysql import db_session
from libraries.bar_functions import do_thing_to_bar
from models.mysql.bar import Bar
from sqlalchemy.orm import Session


def process_the_bar(bar: bar, mysql_session: Session):
    process_result = list()
    context_aware_session = db_session.get()

    # you can do the same in your other functions

```

**Reply by krylosov-aa** [reply]
Hi! I wrote a library that uses context to help manage sessions: https://github.com/krylosov-aa/context-async-sqlalchemy

**Comment by Maxpowerfreak** [comment]
Hey @insomnes 

Seems like exactly what I need! I'll read up on ContextVar a bit more, but at first glance it looks like it covers my case very well.

I'll close this question as it has been answered. Thanks a lot for your help! :smile:

**Comment by cmflynn** [comment]
can you explain where the `db_session` value comes from?

```
db_session: ContextVar[Session] = ContextVar('db_session')
```

Nvm, I missed this line in the handler which actually sets the value of the contextVar:

`db_session.set(mysql_session)`

**Comment by acidjunk** [comment]
I cooked up something based on a bigger project in which I was involved for a customer that needs a lot of DB functionality. It makes the DB available after a simple import. It does use a little bit of postgresql dialect (UUID's for primary id's) but it's mostly normal SQLAlchemy code.

https://github.com/acidjunk/fastapi-postgres-boilerplate

It's a work in progress and it's covered by a blogpost: https://www.renedohmen.nl/deploy-fastapi-on-amazon-serverless/

Hope it helps

**Reply by theobouwman** [reply]
Thanks. I used your implementation and it works really well.

**Reply by ahmedsadman** [reply]
You're using `scoped_session` in conjuction with `contextvars`. What's the point of doing this? Why not just use `sessionmaker` as shown in one of the previous [replies](https://github.com/fastapi/fastapi/discussions/6889#discussioncomment-5132510). Also, if I'm not wrong, `scoped_session` was specifically designed for threaded application, not async

**Reply by acidjunk** [reply]
The DB stuff was mostly copied from a project that does async + threaded stuff.
So there is a FastAPI part and another part that uses a threadpool to execute jobs that also needs to have a DB connection.

Have a look here for a better documented example: https://github.com/workfloworchestrator/orchestrator-core/blob/main/orchestrator/db/database.py

Not sure about the solution, an sich: I only participated somewhat in it, 4 years ago.

https://stackoverflow.com/questions/6519546/scoped-sessionsessionmaker-or-plain-sessionmaker-in-sqlalchemy

**Comment by kenneth-wangg** [comment]
Hey all, is there a downside to importing `get_db` in the file that requires the db connection? In this case, I’m proposing to include these lines in the file containing the `services/mysql/process_bar.py` file

```python
...
from databases.mysql import get_db
...
db = get_db()
```

Any thoughts are appreciated! Thanks.

**Reply by ryanisn** [reply]
I think the downside is you won't have the actual service class with DB calls managed under session context

**Comment by harismuzaffer** [comment]
> 

Why don't just have a global `scoped_session` that returns a new session for every new process. What is the gain by using `ContexVars` over `scoped_session`?

**Reply by ildargafarov** [reply]
Because [scoped_session](https://docs.sqlalchemy.org/en/20/orm/contextual.html#sqlalchemy.orm.scoped_session) uses [threading.local()](https://docs.python.org/3/library/threading.html#threading.local) that isn't compatible with async environment

Edit: There is [async_scoped_session](https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.async_scoped_session) version

**Reply by theobouwman** [reply]
@ildargafarov I think I have a similar issue (https://github.com/tiangolo/fastapi/discussions/10622). Problem is, each db call a new session gets created even when i am using `async_scoped_session`

**Comment by garg10may** [comment]
can't I do something like 

```
class BotCRUD:
    def __init__(self, db: Session= Depends(get_db)):
        self.db = db

    def get_bot(self, bot_id: int):
        return self.db.query(models.Bot).filter(models.Bot.bot_id == bot_id).first()

    def get_bots(self, skip: int = 0, limit: int = 100):
        return self.db.query(models.Bot).offset(skip).limit(limit).all()

    def create_bot(self, bot: BotCreate):
        db_bot = models.Bot(**bot.dict())
        self.db.add(db_bot)
        self.db.commit()
        self.db.refresh(db_bot)
        return db_bot
```

**Comment by ryanisn** [comment]
I'd like to retrieve session from dependency in a non-router module like what @garg10may was asking for, sadly that is not supported in fastapi.
IMO, to truly separate concerns, router should not worry about how db session is composed, it should simply break down requests and forward to responsible service models, handle the result and give it back to response models.
The service model which talks to db models, should be able to make it's own db session decision, whether it is to connect to a fast read replica if this is a read only model, or a special isolation level if there is frequent blocking queries for reporting, etc
And I would like to inject this dependency easily into the service model.
This structure would also be more inline with sqlAlchemy recommendation

> Make sure you have a clear notion of where transactions begin and end, and keep transactions short, meaning, they end at the series of a sequence of operations, instead of being held open indefinitely.

https://docs.sqlalchemy.org/en/20/orm/session_basics.html#:~:text=keep%20transactions%20short

Also there should be no concern with opening a session in each service model, because session will be reused

> By default, when the [Session](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session) is closed, it is essentially in the original state as when it was first constructed, and may be used again

I'm not new to programming but I haven't done professional python for long, let me know if I had anything wrong, thanks

**Comment by jaymeklein** [comment]
I created a decorator which wraps a function/method with the session context:
```python
from functools import wraps
from sqlalchemy.orm import Session
from db.engine import sql_session # returns the ContextVar with the session
def with_session(func):
    """Decorator to inject a session within a class method."""
    
    @wraps(func)
    def wrapper(cls, *args, **kwargs):
    with sql_session() as session:
	    return func(cls, session, *args, **kwargs)
    
    return wrapper
```

So whenever i want to use a context-based session, i do this:
```python
@with_session
def get_all_tenants(cls, session: Session) -> list[Type[Tenant]]:
    return (
	session
	.query(Tenant)
	.filter(Tenant.status)
	.all()
    )
```
The session will be closed when the function/method is finished executing

The ContextVar has been set on the FastAPI lifespan:
```python
@asynccontextmanager
async def lifespan(app: FastAPI) -> AsyncGenerator:
    mysql_session = start_mysql_connection()
    sql_session.set(mysql_session)
    yield
    mysql_session.close()
```

Is this a bad idea?

**Comment by krylosov-aa** [comment]
Hi! Maybe my approach to working with sessions will be useful. It's done through context without dependencies. I'm currently using it with my FastApi app running Postgres.

https://github.com/krylosov-aa/context-async-sqlalchemy
